Nom Etudiante 1: <font color='green'>Ben Mimoun Sarah</font>
<br>
Nom Etudiante 2: <font color='blue'>Kouidrat Shyrin</font>
<br>
Nom Etudiante 3: <font color='red'>Abunada Dima</font>

# TME 2.1 : Projet Detection de motifs - Hash Table

Nous allons développer des algorithmes pour chercher des motifs dans des séquences d'ADN. 

Pour commencer, nous allons générer des données artificielles qui nous permettront de tester rapidement nos algorithmes (**Partie A**). 
</br>
Ensuite, nous allons chercher les motifs dans le génome _C. glabrata_ et analyser les résultats (**Partie B**). 
</br>
Finalement, nous allons appliquer les algorithmes developées pour chercher des motifs dans les données de ChipSeq de _C. glabrata_ (**Partie C**).

## Partie A : Recherche de pattern (motifs) identiques

1\. Écrivez une fonction pour générer aléatoirement des séquences artificielles. Utilisez les fonctions de la bibliothèque `random`. Toutes les nuclétodies sont équiprobables pour la génération des séquences. Si le nombre de séquences est égal à 1 (`n = 1`), renvoyez une chaîne (string) de caractères au lieu d'une liste.

In [1]:
import random

nuc = ('A', 'C', 'G', 'T')

# Pour faire un test simple
n = 3 # nombre de séquences
t = 8 # taille des séquences

def generateRandomSequences(n:int, t:int, upper = True):
    """
    Génére plusieurs séquences nucléotidiques aléatoires 
    entrée n : nombre de séquences
    entrée t : taille des séquences
    entrée upper : bool, si True, les nucléotides seront majuscule, False, minuscule
    sortie séquences : liste de séquences nucléotidiques aléatoires ou un string si n = 1
    """
    sequences = [] # liste des séquences nucléotidiques 
    for i in range(n) :
        sequence = "" 
        for j in range(t) : 
            # ajoute t nucléotides aléatoirement
            sequence += nuc[random.randint(0, 3)]
        sequences.append(sequence)
    return sequences

generateRandomSequences(n = n, t = t, upper = True)

['ACGCGGAG', 'CGGGCATC', 'TTAACCGC']

In [2]:
type(generateRandomSequences(n = 1, t = t, upper = True))

list

2\. Puis, générez `n = 10` séquences de `t = 41` nucléotides chacune. Pour obtenir le même résultat dans le développement de ce TME, choisissez un "seed", mais différent de celui de vos camarades. Pensez à changer le `seed` pour effectuer des tests variés.

In [3]:
# Pour obtenir les séquences artificielles
n = 10
t = 40

seed = 77

random.seed(seed)

seq_artificielles_sans_motif = generateRandomSequences(10,41)

for seq in seq_artificielles_sans_motif :
    print(seq)

GGCCCAGTCCAGATCCTCGGAAGTCCCAACTAACGAATAAG
TAGATCCTTCTAAATAGTAGTATACGAATGTAATCAACATA
TTACCGCACCCAACTTTGATTGGGTCATAAACAAACATCAT
GTTTCTGGAGGCGGTTTGTGAGACCGACAAACGCGATAGAG
TAATCTTATAAATCAGGCAAATACTCGGTCCTTTCTACGGG
TCGCGGCCGCGTGGTCATAGGCAGTTTTATGTTTCTTTGAG
GTTAATGGGAGGCATCCATGGGAGCCGTATATCCATATAAA
AGTTGTACGCACCCGGGAACGTTCATTTTTAGCATCGGGGT
GGCAAACATACTCTCGGCAGTTCTTTCCAAACAGTGTAGTA
TACGTCACGGAGGACGGTCTGGCATCAATTCTCAGAGGCTC


3\. Nous allons maintenant implanter dans nos séquences artificielles un motif de taille `k = 4` à des positions aléatoires (choisies uniformément le long de la séquence). **BONUS :** prise en compte du paramètre `f` qui permet de contrôler le pourcentage de séquences qui contiennent le motif. Par exemple, `n = 10` et `f = 0.9` signifie que 9 séquences sur 10 contiennent le motif. ` 

In [4]:
?random.sample 
# fonction qu'on utilise dans implantMotifs

Signature: random.sample(population, k, *, counts=None)
Docstring:
Chooses k unique random elements from a population sequence.

Returns a new list containing elements from the population while
leaving the original population unchanged.  The resulting list is
in selection order so that all sub-slices will also be valid random
samples.  This allows raffle winners (the sample) to be partitioned
into grand prize and second place winners (the subslices).

Members of the population need not be hashable or unique.  If the
population contains repeats, then each occurrence is a possible
selection in the sample.

Repeated elements can be specified one at a time or with the optional
counts parameter.  For example:

    sample(['red', 'blue'], counts=[4, 2], k=5)

is equivalent to:

    sample(['red', 'red', 'red', 'red', 'blue', 'blue'], k=5)

To choose a sample from a range of integers, use range() for the
population argument.  This is especially fast and space efficient
for sampling from a lar

In [5]:
k = 4 # taille des motifs

# Pour le bonus
f = 0.9 # fréquence d'occurences des motifs

def implantMotifs(motif:str, sequences:list, f = 0.9):
    """
    Insére un motif à des positions aléatoires des séquences
    entrée motif : motif qui va être implanté dans les séquences
    entrée séquences : liste de séquences
    entrée f : fraction des séquences qui contiendront le motif
    sortie modified_sequences: liste de séquences ayant le motif implanté
    """
    n = len(sequences) # nombre de séquences 
    nb_implant = int(f * n) # nombre de séquences comportant le motif 
    # Copie des séquences pour les modifier
    modified_sequences = sequences.copy()
    # choix UNIQUE des indices à modifier (il n'y a pas de doublon avec cette fonction)
    indices_modifies = random.sample(population = range(n), k = nb_implant) # pour n séquences en choisir aléatoirement f*n
    
    for i in indices_modifies :
        sequence = modified_sequences[i] # séquence à modifier 
        pos = random.randint(0, len(sequence)) # position uniforme
        modified_sequences[i] = sequence[:pos] + motif + sequence[pos:] # insertion 

    return modified_sequences

# Pour obtenir le même résultat que dans le dévelopent de ce TME
random.seed(seed)

motif_artificiel: str =  generateRandomSequences(1, k)[0]
seq_artificielles =  generateRandomSequences(10, 41)

print("Le motif géneré est : ", motif_artificiel)

print("\nSéquences artificielles :")
for seq in seq_artificielles :
    print(seq)
    
print("\nSéquences modifiées :")
for seq in implantMotifs(motif_artificiel, seq_artificielles, f) : 
    print(seq)

Le motif géneré est :  GGCC

Séquences artificielles :
CAGTCCAGATCCTCGGAAGTCCCAACTAACGAATAAGTAGA
TCCTTCTAAATAGTAGTATACGAATGTAATCAACATATTAC
CGCACCCAACTTTGATTGGGTCATAAACAAACATCATGTTT
CTGGAGGCGGTTTGTGAGACCGACAAACGCGATAGAGTAAT
CTTATAAATCAGGCAAATACTCGGTCCTTTCTACGGGTCGC
GGCCGCGTGGTCATAGGCAGTTTTATGTTTCTTTGAGGTTA
ATGGGAGGCATCCATGGGAGCCGTATATCCATATAAAAGTT
GTACGCACCCGGGAACGTTCATTTTTAGCATCGGGGTGGCA
AACATACTCTCGGCAGTTCTTTCCAAACAGTGTAGTATACG
TCACGGAGGACGGTCTGGCATCAATTCTCAGAGGCTCGTAC

Séquences modifiées :
CAGTCCAGATCCTCGGAAGTCCCAACTAACGAATAAGTAGGCCGA
TCCTTCTAGGCCAATAGTAGTATACGAATGTAATCAACATATTAC
CGCACCCAAGGCCCTTTGATTGGGTCATAAACAAACATCATGTTT
CTGGAGGCGGTTTGTGAGACCGACAAACGCGGGCCATAGAGTAAT
CTTATAAATCAGGCAAATACTCGGTCCTTTCTACGGGTCGC
GGCCGCGTGGTCATAGGCAGTTTTATGTTTCTTTGAGGGGCCTTA
ATGGGAGGCATCCATGGGAGCGGCCCGTATATCCATATAAAAGTT
GTACGGGCCCACCCGGGAACGTTCATTTTTAGCATCGGGGTGGCA
AACATACTCTCGGCAGTTCTTTCCAAACAGTGTGGCCAGTATACG
TCACGGAGGACGGTCGGCCTGGCATCAATTCTCAGAGGCTCGTAC


4\. Faites une fonction pour chercher les $m$ motifs de taille $k$ les plus fréquents dans l'ensemble des séquences.

In [6]:
def searchMotifs(k:int, sequences:list):
    """
    Cherche les motifs de taille k dans un ensemble de séquences
    entrée k : taille du motif
    entrée séquences : liste de séquences
    sortie motifs: dictionnaire de motifs, clé = motif, valeur = fréquence d'observation
    >>>searchMotifs(3, ['TAAGTAA', 'TATAA', 'CTATC'])
    {'TAA': 3, 'AAG': 1, 'AGT': 1, 'GTA': 1, 'TAT': 2, 'ATA': 1, 'CTA': 1, 'ATC': 1}
    """
    motifs  = {} # dictionnaire des motifs (motif : occurence)
    for seq in sequences :
        for i in range(len(seq) - k + 1) : 
            motif = seq[i : i + k]
            if motif in motifs :
                motifs[motif] += 1
            else:
                motifs[motif] = 1
    return motifs

test_motifs = searchMotifs(3, ['TAAGTAA', 'TATAA', 'CTATC'])

print(test_motifs)

{'TAA': 3, 'AAG': 1, 'AGT': 1, 'GTA': 1, 'TAT': 2, 'ATA': 1, 'CTA': 1, 'ATC': 1}


5\. Faite aussi une fonction qui affiche les `top` motifs les plus fréquents. Cette fonction peut également trier le dictionnaire de manière croissant ou décroissant.  
*Note*: Voir les paramètres `key` et `reverse` de la fonction `sorted`

In [7]:
?sorted

Signature: sorted(iterable, /, *, key=None, reverse=False)
Docstring:
Return a new list containing all items from the iterable in ascending order.

A custom key function can be supplied to customize the sort order, and the
reverse flag can be set to request the result in descending order.
Type:      builtin_function_or_method

In [8]:
def getTopMotifs(motifs:dict, top, decroissant = True):
    """
    Renvoie les top motifs le plus fréquent
    entrée motifs: dictionnaire de motifs, clé = motif, valeur = fréquence d'observation
    entrée top : les top plus fréquent, si top == "all", renvoie le dictionnaire complet trié sans filtrer
    entrée decroissant : bool, si True le dictionnaire est trié par ordre décroissant de valeur
    sortie motifsfreq: dictionnaire contenant les top motifs les plus fréquents, clé = motif, valeur = fréquence d'observation
    
    >>>getTopMotifs({'TAA': 3, 'AAG': 1, 'AGT': 1, 'GTA': 1, 'TAT': 2, 'ATA': 1, 'CTA': 1, 'ATC': 1}, 2)
    {'TAA': 3, 'TAT': 2}
    """
    # Tri du dictionnaire par fréquence en retournant une liste :
        # liste tuples (cles,val)
        # sur quoi trier : ici les valeurs donc le deuxième élèment du tuple (fonction anonyme)
        # en ordre décroissant (ou pas)
    sorted_motifs = sorted(motifs.items(), key = lambda x: x[1], reverse =  decroissant)

    if top != "all" : 
        # Filtrer selon le nombre top demandé (déjà triée donc les 1er)
        sorted_motifs = sorted_motifs[:top]
    
    # Reconstruire en dictionnaire
    motifsfreq = dict(sorted_motifs)

    return motifsfreq

    
test_motifs_top = getTopMotifs(test_motifs, 2, decroissant = True)
    
print(test_motifs_top)

{'TAA': 3, 'TAT': 2}


6\. Testez cette fonction sur vos séquences articielles avec le motif implanté généré dans les questions précédentes. Faites attention à ne pas comparer des séquences en majuscules avec des minuscules. Avez-vous trouvé le motif implanté ? Si oui, à quelle fréquence ?

In [9]:

motif_artificiel: str =  generateRandomSequences(1, k)[0]
seq_artificielles =  generateRandomSequences(10, 41)
print("Le motif géneré est : ", motif_artificiel)

seq_motif = implantMotifs(motif_artificiel, seq_artificielles)
occurences_motif_articiel = searchMotifs(k, seq_motif)
occurences_motif_articiel_top = getTopMotifs(occurences_motif_articiel, 1) 

print("motif le plus courant : ", occurences_motif_articiel_top)
print("Fréquence observé : ", occurences_motif_articiel[motif_artificiel])

Le motif géneré est :  CTTT
motif le plus courant :  {'CTTT': 12}
Fréquence observé :  12


    -> On a trouvé le motif implanté, 11 pour fréquence. 

7\. Quelle est la complexité de votre algorithme?

    searchmotif : len(sequences)*k*len(seq) 

    gettopmotif  ? 

8\. Certains motifs sont reverse complémentaires. Implantez ces motifs reverse complémentaires dans de nouvelles séquences artificielles et regroupez les deux listes de séquences, puis recherchez-y ces motifs. **Attention** : vous devez réutiliser les fonctions développées précédemment. Il n’est pas nécessaire de définir de nouvelles fonctions.

Pour cela, nous allons développer un module **Python** appelé **motifsSearch**.

* Créez un fichier nommé `motifsSearch.py`.
* Copiez et collez les fonctions déjà développées dans ce fichier texte.
* Sauvegardez-le sous ce nom dans le même répertoire que les fichiers **Jupyter** où vous allez l'utiliser.

In [10]:
import os
for f in os.listdir():
    print(repr(f))


'seqs_artificielles_avec_motif_et_rv.txt'
'__pycache__'
'.ipynb_checkpoints'
'seqs_artificielles_sans_motif.txt'
'motif_artificiel.txt'
'TME2.1-HashTable_2026.ipynb'
'seqs_artificielles_avec_motif.txt'
'motifsSearch.py'
'C_glabrata_1000bp_only.fasta'


In [11]:
# Reutiliser la fonction déjà defini (TME1.2_ORFs, Question 3)
from motifsSearch import reversecompl

motif_rv = reversecompl(motif_artificiel)

random.seed(seed)
seq_artificielles_avec_rv = list(map(reversecompl, seq_artificielles))

occurences_motif_artificiel = searchMotifs(k, seq_artificielles_avec_rv)

occurences_motif_artificiel_top = getTopMotifs(occurences_motif_articiel, 1) 

print("Le motif original est : ", motif_artificiel)
print("Le motif reverse est : ", motif_rv)
print(occurences_motif_articiel_top)

TGCCACGTT
Le motif original est :  CTTT
Le motif reverse est :  AAAG
{'CTTT': 12}


9\. Sauvegardez vos séquences artificielles avec et sans le motif ainsi que le motif généré pour les reutiliser les séances suivantes.

In [12]:
seqs_file = "seqs_artificielles_sans_motif.txt"

seqs_motif_file = "seqs_artificielles_avec_motif.txt"

seqs_motif_rv_file = "seqs_artificielles_avec_motif_et_rv.txt"

motif_file = "motif_artificiel.txt"


In [13]:
fichier = open(seqs_file, "a")
fichier.write("".join(seq_artificielles))
fichier.close()

In [14]:
fichier = open(seqs_motif_file, "a")
fichier.write("".join(seq_motif))
fichier.close()

In [15]:
fichier = open(seqs_motif_rv_file, "a")
fichier.write("".join(list(map(reversecompl, seq_motif))))
fichier.close()

In [16]:
fichier = open(motif_file, "a")
fichier.write("".join(list(map(reversecompl, motif_artificiel))))
fichier.close()


## Partie B : Recherche de motifs identiques sur le génome de _C. glabrata_

1\. Le fichier **C_glabrata_1000bp_only.fasta** contiens les séquences régulatrices pour tous les gènes de votre organisme. Nous avons pris les 1000bp en amont du codon start. Cherchez les 50 motifs de taille 7 les plus fréquents dans votre génome. Pensez aussi à mesurer le temp utilisé pour obtenir les résultats avec la fonction `time.time()`.

In [17]:
import time

k = 7
top = 50

def readFasta(fastaFileName):
    """
    Lit un fichier fasta
    entrée fastaFileName: nom du fichier fasta
    sortie séquences: liste contenant toutes les séquences du fichier
    """
    
    sequence = ""
    sequences_list = []
    prev_header = ""
    header = ""
        
    for line in open(fastaFileName):
        string = line.strip()
        if string[0] != ">":
            if prev_header != header:
                prev_header = header
            sequence = sequence + string
        else:
            header = string
            if sequence != "":
                sequences_list.append(sequence)
                sequence = ""

    sequences_list.append(sequence)
    return sequences_list

# Lire le génome - renvoie une liste de str
genome = readFasta("C_glabrata_1000bp_only.fasta")

#  Mesurez le temps de calcul
start = time.time()

genome_motifs =  searchMotifs(k = 7, sequences = genome)
genome_motifs_top = getTopMotifs(genome_motifs, top = 50)

end = time.time()

print("Top motifs :\n", genome_motifs_top)

print("Temp (secondes) :", end - start)

Top motifs :
 {'AAAAAAA': 8473, 'TTTTTTT': 7684, 'ATATATA': 3117, 'AAAAAAT': 2990, 'ATTTTTT': 2983, 'TATATAT': 2809, 'AAAAATT': 2487, 'AATTTTT': 2464, 'AAAGAAA': 2449, 'GAAAAAA': 2430, 'AAAATTT': 2326, 'AAATTTT': 2317, 'TATTTTT': 2303, 'TTTTTTC': 2294, 'AAGAAAA': 2279, 'AAAAATA': 2220, 'TATATAA': 2211, 'TTTATTT': 2174, 'TTTCTTT': 2167, 'TGAAAAA': 2158, 'TTATTTT': 2145, 'TTTTATT': 2117, 'AAAAGAA': 2104, 'TTTTCTT': 2057, 'CAAAAAA': 1991, 'TTTTTCA': 1980, 'ATATAAA': 1972, 'AAAATAA': 1964, 'ATATATT': 1921, 'AAATAAA': 1918, 'AATAAAA': 1918, 'TTCTTTT': 1915, 'AATATAT': 1914, 'AGAAAAA': 1885, 'AAAAAAG': 1878, 'TTTTTAT': 1873, 'CTTTTTT': 1869, 'AATAATA': 1866, 'AAATATA': 1835, 'TTTTTTA': 1821, 'TTTTTTG': 1818, 'ATAAAAA': 1813, 'TTTTTCT': 1808, 'ATATTTT': 1792, 'TATATTT': 1779, 'TTATATA': 1749, 'TATTATT': 1738, 'AAAATAT': 1718, 'AAAAAGA': 1698, 'TAAAAAA': 1675}
Temp (secondes) : 2.5447874069213867


2\. Quel sont les trois motifs de taille 7 les plus fréquents? Pensez-vous que ces motifs correspondent à de facteur de transcription connus? Justifiez votre réponse.

<font color='blue'>
'AAAAAAA': 8473, 'TTTTTTT': 7684, 'ATATATA': 3117
ce n'est rien du tout, homopolymère, motifs trop simples.
</font>

3\. Le motif **TGATTCAT** correspond au facteur de transcription **Gcn4**, qui est très fréquemment retrouvé dans le génome des levures. Développez une fonction pour rechercher un motif spécifique dans un dictionnaire de motifs trouvés.

In [18]:
print(test_motifs)

{'TAA': 3, 'AAG': 1, 'AGT': 1, 'GTA': 1, 'TAT': 2, 'ATA': 1, 'CTA': 1, 'ATC': 1}


In [28]:
def searchGivenMotif(motifsDict, motifSpecifique, decroissant = True):
    """
    Cherche un motif specifique dans un dictionnaire de motifs trouvés
    entrée motifsDict : dictionnaire de motifs, clé = motif, valeur = fréquence d'observation
    entrée motifSpecifique: un motif specifique à chercher
    entrée decroissant : bool, si True, le dictionnaire est trié par ordre décroissant de valeur
    sortie fréquence : la fréquence du motif
    sortie ranking : dans quelle position le motif a été trouvé
    >>>searchGivenMotif(test_motifs, "TAT")
    (2, 2)
    """
    motifsList = sorted(motifsDict.items(), key = lambda x: x[1], reverse = decroissant)
    # motifsList est une liste de tuples (motif, freqc)

    resultat = next((t for t in motifsList if motifSpecifique in t[0]), None) # le tuple qui est de la forme (motifSpecifique,_)
    if resultat == None : # si le motif n'est pas dans le dictionnaire 
        print(motifSpecifique + "N'est pas présent")
        return 0, 0
        
    index = motifsList.index(resultat)
        
    ranking = index + 1 # le rang commence à 0
    frequence = motifsList[index][1]
    
    return ranking, frequence
 
searchGivenMotif(test_motifs, "TAT", decroissant = True)

(2, 2)

4\. Essayez votre fonction avec le dictionnaire de motifs de _C. glabrata_. Est-ce que vous avez trouvé le motif de **Gcn4** ? Si oui, à quel ranking et avec quelle fréquence ?

In [30]:
motif_Gcn4 = 'TGATTCA'

gcn_rank, gcn_freq = searchGivenMotif(genome_motifs, motif_Gcn4, decroissant = True)

print("Motif Gcn4 trouvé\nRanking : %s\nFréquence : %s" % (gcn_rank, gcn_freq))

Motif Gcn4 trouvé
Ranking : 1834
Fréquence : 593


Oui nous l'avons trouvé rang : 1834 et à une fréquence 593. 

5\. Les motifs peu complexes, avec par exemple 5, 6 ou 7 fois la même lettre (homopolymères), sont courants dans les génomes et ils n'ont généralement pas de signification biologique. Faites une fonction pour éliminer les motifs peu complexes, _i. e._ qui ont au moins $m$ fois la même lettre. Faites attention au type d'objet d'entrée pour chaque fonction.

In [ ]:
def removeLowComplexeHomo(motifs:list, m:int):
    """
    Enlève les motifs peu complexe ayant m fois le même nucléotide
    entrée motifs: liste de motifs
    entrée m: nombre de nucleotides répétés dans un motif peu complexe
    sortie motifsClean: liste de motifs sans les motifs peu complexe
    >>>removeLowComplexeHomo(['TAA', 'AAG', 'AGT', 'GTA', 'TAT', 'ATA', 'CTA', 'ATC'], 2)
    ['AGT', 'GTA', 'CTA', 'ATC']
    """

    motifsClean = []

    return motifsClean

test_motifs_list # Code

removeLowComplexeHomo(test_motifs_list, 2)

6\. **BONUS :** Dans le context de ce projet, vous pouvez eliminer aussi les motifs constitués d'un groupe de deux lettre répétées, comme par exemple **AGAGAGA** (heteropolymères). Vous pouvez aussi penser à supprimer les heteropolymères avec une variation, par example **AGAGTAG**. Créez une fonction pour enlever ce type de motifs peu complex dans le dictionnaire test ci-dessous. Réfléchissez aux meilleurs paramètres pour reproduire le résultat exemple dans la description de la fonction.

In [ ]:
import re # TME1.2

def removeLowComplexeHetero(motifs:list, n:int, variation = "yes"):
    """
    Enlève les motifs peu complexe ayant n fois un dinucléotide
    entrée motifs: liste de motifs, clé = motif, valeur = fréquence d'observation
    entrée n: nombre di-nucleotides répétés dans un motif peu complexe
    entrée variation : string, si "yes" permettre variation d'un nucléotide
    sortie motifsClean: liste de motifs sans les motifs peu complexe
    >>>removeLowComplexeHetero(['GGTTTGG', 'TGAGTTA', 'TGCCGTG', 'AGAGAGA', 'TCACCGA', 'TTGGTAT', 'AGGGTGG', 'TGGCTTA', 'AGAGTAG', 'GCCCCTC'], 3, variation = "yes")
    ['GGTTTGG', 'TGAGTTA', 'TGCCGTG', 'TCACCGA', 'TTGGTAT', 'TGGCTTA']
    """
    
    motifsClean = []
    return motifsClean

Tests et résultats attendus.

In [ ]:
m = 5 # 5 fois le meme repete: trop simple

motifs_to_filter = ['GGTTTGG', 'TGAGTTA', 'TGCCGTG', 'AGAGAGA', 'TCACCGA', 'TTGGTAT', 'AGGGTGG', 'TGGCTTA', 'AGAGTAG', 'GCCCCTC']

motifs_filtered_homo = # Code

# Si obtenu
motifs_filtered_homo_hetero_novar_3 = # n=3, variation=False
motifs_filtered_homo_hetero_var_3 = # n=3, variation=True
motifs_filtered_homo_hetero_var_2 = # n=2, variation=True

print(motifs_filtered_homo)
# Si obtenu
print(motifs_filtered_homo_hetero_novar_3)
print(motifs_filtered_homo_hetero_var_3)
print(motifs_filtered_homo_hetero_var_2)


In [ ]:
# Attention au type d'objet

genome_motifs_list = genome_motifs.keys()# Code
print("K-mers : %s" % (len(genome_motifs)))

genome_motifs_list_filtree = removeLowComplexeHomo(genome_motifs_list, 5) # Code
print("K-mers sans homopolymères : %s" % (len(genome_motifs_list_filtree)))

# BONUS
#genome_motifs_list # Code, si obtenu
#print("K-mers sans heteropolymères (variation = 'no') : %s" % (len(genome_motifs_list)))

# BONUS
#genome_motifs_list # Code, si obtenu
#print("K-mers sans heteropolymères (variation = 'yes') : %s" % (len(genome_motifs_list)))

8\. Une fois que vous avez appris que les motifs peu complexes mènent à recherches inutiles, créez une fonction `hashTable`, version ameliorée de `searchMotifs()`, qui prend comme entrée une liste de séquences et une liste de motifs de longueur `k` filtrée (sans peu complexes). Cette fonction considère seulement les motifs dans la liste donnée.

In [ ]:
def hashTable(sequences:list, kmersV:list):
    """
    Cherche les motifs de taille k dans un ensemble de séquences avec la méthode de Hash Table
    entrée séquences : liste de séquences
    entrée kmersV: liste de Kmers valides à chercher
    sortie motifs_freq_dict: dictionnaire de motifs, clé = motif, valeur = fréquence d'observation
    >>>hashTable(['TAAGTAA', 'TATAA', 'CTATC'], ['TAA', 'AAG', 'AGT', 'GTA', 'TAT', 'ATA', 'CTA', 'ATC'])
    {'TAA': 3, 'AAG': 1, 'AGT': 1, 'GTA': 1, 'TAT': 2, 'ATA': 1, 'CTA': 1, 'ATC': 1}
    """
    
    motifs_freq_dict  = {}
    return motifs_freq_dict

hashTable(['TAAGTAA', 'TATAA', 'CTATC'], ['TAA', 'AAG', 'AGT', 'GTA', 'TAT', 'ATA', 'CTA', 'ATC'])

9\. Maintenant, cherchez les motifs de taille 7 les plus fréquents dans les deux brin de votre génome. Utilisez la liste de motifs (k-mers) créée precedement.
Affichez le résultat pour les 20 premiers motifs: est-il le même dans les deux brins ? Dans quel position vous trouvez le motif **Gcn4** en utilisant la fonction `hashTable()` (considérez les deux brins) ? Justifiez votre réponse.

**Brin 1**

In [ ]:
genome_motifs # Code

genome_motifs_top # Code

print("Top motifs :\n", genome_motifs_top)
# Afficher seulement les 20 premiers 

**Brin 2**

In [ ]:
# Générer les séquences reverse complement
genome_rev # Code

genome_motifs_rev # Code

genome_motifs_rev_top # Code

print("Top motifs :\n", genome_motifs_rev_top)
# Afficher seulement les 20 premiers 

In [ ]:
# Affiche résultats
print("Brin 1")
gcn_rank_filt, gcn_freq_filt = # Code
print("Motif Gcn4 trouvé\nRanking : %d\nFrequence : %d" % (gcn_rank_filt, gcn_freq_filt))

print("\n")

print("Brin 2")
gcn_rank_filt, gcn_freq_filt = # Code
print("Motif Gcn4 trouvé\nRanking : %d\nFrequence : %d" % (gcn_rank_filt, gcn_freq_filt))

<font color='blue'>
Votre réponse
</font>

10\. **BONUS**: Développez une fonction qui, à partir d'un dictionnaire de motifs, regroupe les résultats du motif reverse complémentaire afin de faciliter la comparaison. Le dictionnaire de sortie doit 
- avoir pour clé un tuple `(motif, motif_rv)` 
- pour valeur un tuple `(fréquence fw, frequence rv)` 
- être trié de sorte que les motifs dont la somme `(fréquence fw + frequence rv)` est plus grande apparaissent en premier. 

In [ ]:
def gatherRevCompMotifs(motifs_dict):
    """
    Trouver les motifs reverse complementaire et rassembler leur résultats
    entrée motifs_dict : dictionaire de résultats de recherche de motifs
    sortie motifs_dict_gather : dictionaire de motifs, clé = motif, valeur = valeur motif fw, valeur motif rv
    """
    
    motifs_dict_gather = {}
    return motifs_dict_gather

genome_motifs_gat = # Code

print("Dictionnare rassemblé:")
# Afficher les 10 premiers éléments

# Rang du motif Gcn4 dans le dictionnaire rassemblé
gcn_rank_filt, gcn_freq_filt = searchGivenMotif(genome_motifs_gat, motif_Gcn4)# Code
print("Motif Gcn4 trouvé\nRanking : %d\nFrequence : %s" % (gcn_rank_filt, gcn_freq_filt))

## Partie C : Recherche de motifs identiques sur vos données

1\. Le fichier "Sequence_by_Peaks_##.fasta" contiens les regions de peak trouvé par ChipSeq, qui contient probablement le Facteur de Transcription que nous cherchons. Apliquez les fonctions precedents pour chercher les 3 motifs les plus fréquents dans les deux brim. Il faut bien evidement enlever les motifs peu complexe. Essayez plusier tailles de motifs.

In [ ]:
k = 7 # tester d'autres tailles k = 6, k = 8, etc.
m = 5 # tester aussi d'autres valeurs
n = 2 # 3 ? 
variation = True # False ? 

peaks = # Code

# Code: liste de motifs, filter ... 

peaks_motifs_gat = # Code

print("Top motifs :\n", peaks_motifs_gat)

2\. Ulitsez la base YEASTRACT : http://www.yeastract.com/formsearchbydnamotif.php pour chercher les motifs. 
Avez vous une indication pour le facteur de transcription impliqué ?

<font color='blue'>
Votre réponse
</font>